contracts/AckToEarn.sol

```diff
@@ -0,0 +1,124 @@
+        uint amount;
+        uint timestamp;
+        string message;
+        bool claimed;
+        bool exists;
+    }
+
+    mapping(address => mapping(uint => Bid)) private bidderBids;
+    mapping(address => mapping(uint => Bid)) private recipientBids;
+
+    Bid[] public bids;
+
+    uint constant bidExpiryThreshold = 7 days;
+
+    /**
+    * @notice Send a message and bid amount to a recipient
+    */
+    function sendBid(string memory message, address recipient) external payable {
+        require(recipient != address(0), "Recipient cannot be the zero address");
+
+        // The recipient gets 90% of the bid amount
+        uint recipientAmount = (msg.value * 90) / 100;
+        uint newBidId = bidIds.current();
+        Bid memory bid = Bid({
+            id: newBidId,
+            bidder: msg.sender,
+            recipient: recipient,
+            amount: msg.value,
+            message: message,
+            timestamp: block.timestamp,
+            claimed: false,
+            exists: true
+        });
+
+        bids.push(bid);
+
+        bidderBids[msg.sender][newBidId] = bid;
+        recipientBids[recipient][newBidId] = bid;
+
+        emit NewBid(msg.sender, recipient, msg.value);
+    }
+
+    /**
+    * @notice Returns all bids that have ever been made on the platform
+    */
+    function getBids() external view returns(Bid[] memory) {
+        return bids;
+    }
+
+    /**
+    * @notice Reclaims a specific set of expired bids that the bidder has made.
+    */
+    function reclaimBids(uint[] memory bidIdsToReclaim) external nonReentrant {
+        uint totalReclaimAmount = 0;
+
+        for (uint i = 0; i < bidIdsToReclaim.length; i++)
+        {
+            uint bidId = bidIdsToReclaim[i];
+            Bid memory bid = bidderBids[msg.sender][bidId];
+
+            if (!bid.exists) {
+                continue;
+            }
+            // If the recipient has already claimed the bid, don't allow the balance to be reclaimed
+            if (bid.claimed) {
+                continue;
+            }
+            // If the bid has not expired yet then don't allow the balance to be reclaimed
+            if (!isBidExpired(bidId, msg.sender)) {
+                continue;
+            }
+
+            uint bidReclaimAmount = (bid.amount * 90) / 100;
+            totalReclaimAmount += bidReclaimAmount;
+            bidderBids[msg.sender][bidId].claimed = true;
+        }
+
+        require(totalReclaimAmount > 0, "No ether to reclaim.");
+
+        (bool success, ) = (msg.sender).call{value: totalReclaimAmount}("");
+
+        require(success, "Failed to send ether.");
+
+        emit ReclaimBid(msg.sender, totalReclaimAmount);
+    }
+
+    /**
+    * @notice Check if bid is expired
+    * @dev Returns true if the bid cannot be found
+    */
+    function isBidExpired(uint bidId, address account) public view returns(bool) {
+        bool isBidderAccount = bidderBids[account][bidId].exists == true;
```

Comment:
do we need the `== true`?

Reply:
also, i'm not really sure what `exists` is doing for us. could use more clarity on what we are doing with that

Reply:
Yea, maybe I should add a comment about those checks...

No, we don't need the `== true`, good catch.

So in order to find out if a bid is expired, we need to find it in either the bidder or recipient map. The problem is that there apparently is no direct way to find out if a value exists in a map in Solidity (https://ethereum.stackexchange.com/a/13029). Thus, adding `exists` to the `Bid` structure allows us a cleaner, more readable way to check if a bid exists in the map or not. If a map element does not exist then the default value of `0` is returned. Theoretically, we could use this default value and omit the `exists` parameter. I personally like the exists parameter over using the default value here but I'm open to removing it if it's causing confusion...

Here's another resource: 

https://stackoverflow.com/a/49637782